File: backend/app/schemas/identity.py
"""Clone-risk API contracts.

`explanation` is rendered server-side and shipped as a finished sentence, the
same convention `ScoreReason.message` and the Radar's reasons follow. A client
that composed this string would own a verdict the platform did not issue, and
the two could then disagree about the same token.
"""

from __future__ import annotations

from typing import Literal

from app.schemas.common import BaseSchema


class IdentityOut(BaseSchema):
    """Whether a token's name is contested, and where it sits in the queue."""

    mint_address: str

    #: Tokens sharing this exact name, counting this one. 1 means unique.
    sharing_name: int
    #: How many of those MEMESCOPE discovered before this token.
    discovered_before: int
    #: Earliest *observed*, which is not the same as earliest in existence —
    #: the scanner cannot see launches that predate it.
    is_earliest_known: bool

    clone_risk: Literal["none", "low", "moderate", "high"]
    identity_confidence: Literal["high", "moderate", "low"]

    #: A finished sentence. Display it; do not parse it.
    explanation: str


class IdentityPage(BaseSchema):
    items: list[IdentityOut]

File: frontend/src/types/identity.ts
/**
 * Clone-risk contracts.
 *
 * Mirrors `backend/app/schemas/identity.py`. `explanation` arrives as a
 * finished sentence and is displayed verbatim — the client never composes or
 * parses it, for the same reason it never composes a score reason.
 */

export type CloneRisk = "none" | "low" | "moderate" | "high";
export type IdentityConfidence = "high" | "moderate" | "low";

export interface TokenIdentity {
  mint_address: string;
  /** Tokens sharing this exact name, including this one. 1 means unique. */
  sharing_name: number;
  discovered_before: number;
  /** Earliest *observed*, which is not the same as earliest in existence. */
  is_earliest_known: boolean;
  clone_risk: CloneRisk;
  identity_confidence: IdentityConfidence;
  explanation: string;
}

export interface IdentityPage {
  items: TokenIdentity[];
}
